### PR TITLE
File Handling

### PR DESCRIPTION
**Pull request by mmthornt**
Rubber Duck will inform the user that it cannot read files if a file is submitted to it.
Rubber Duck will add a link to thread created for user

### CODE REVIEW

File: Readme.md
Comment (byubean) [inline Readme.md:1]: Google "windows git keep unix newlines" and follow the instructions to make sure you aren't checking in windows newlines into github.

File: discord_bot.py [hunk 1]
```diff
@@ -1,282 +1,297 @@
...
+                current_fence = ""
+            else:
+                current_fence = line
+
+    if block:
+        yield block
+
+
+def as_message(message: discord.Message) -> Message:
+
```
Comment (byubean) [inline discord_bot.py:61]: Unnecessary empty line.

File: discord_bot.py [hunk 2]
```diff
@@ -1,282 +1,297 @@
...
+        message_id=message.id,
+        
+        content=message.content,
+        # is_file=len(message.attachments) > 0
+
+        file = [as_attachment(attachment) for attachment in message.attachments] # call new as_attachment
+    )
+#create function that takes a attachment and return a dictionary
+#converts each attachment to a dictionary try __dict()
+# get filename and id
```
Comment (byubean) [inline discord_bot.py:78]: Remove TODO comments once completed.

File: discord_bot.py [hunk 3]
```diff
@@ -1,282 +1,297 @@
...
+
```
Comment (byubean) [inline discord_bot.py:79]: Make sure your IDE has a pep8 formatter installed (VSCode and PyCharm both do). Use it before you commit your code.

File: rubber_duck.py [hunk 4]
```diff
@@ -1,156 +1,184 @@
...
+class Message(TypedDict):
+    guild_id: int
+    channel_name: str
+    channel_id: int
+    author_id: int
+    author_name: str
+    author_mention: str
+    message_id: int
+    content: str
+    # swapped
```
Comment (byubean) [inline rubber_duck.py:31]: Remove comment

File: rubber_duck.py [hunk 5]
```diff
@@ -1,156 +1,184 @@
...
+    file: list
+    # is_file: bool
```
Comment (byubean) [inline rubber_duck.py:33]:
```python
file: list[Attachment]
```
and remove commented obsolete code.

File: rubber_duck.py [hunk 6]
```diff
@@ -1,156 +1,184 @@
...
+
+#create class attachment extends type dict and import to discord bot
```
Comment (byubean) [inline rubber_duck.py:35]: Remove comment

File: rubber_duck.py [hunk 7]
```diff
@@ -1,156 +1,184 @@
...
+
+                message_history.append(GPTMessage(role='user', content=message['content']))
+
+                # replaced by recommended lines (102-104) above line 106 (101 originally)
+                # if message['is_file'] == True:
+                #     appended_content = "*The user attached a file*"
+                # else:
+                #     appended_content = message['content']
+
+                # message_history.append(GPTMessage(role='user', content=appended_content))
```
Comment (byubean) [inline rubber_duck.py:121]: Delete obsolete comments

File: rubber_duck.py [hunk 8]
```diff
@@ -1,156 +1,184 @@
...
+                await self._metrics_handler.record_message(
+                    guild_id, thread_id, user_id, message_history[-1]['role'], message_history[-1]['content'])
+
+                try:
+                    if message_history[-1]['content'] == "*The user attached a file*":
+                        response = "I'm unable to read entire files directly. Could you please paste the relevant sections and rephrase your question? " \
+                                   "This will help me assist you better!"
+
+                        await self._metrics_handler.record_message(
+                            guild_id, thread_id, user_id, 'assistant', response)
```
Comment (byubean) [inline rubber_duck.py:135]: What is this for?

It looks extremely brittle. Comparing the last message content to a hard-coded string is bound to break.

Comment (byubean) [reply rubber_duck.py:135]: It looks like a leftover from the code commented out on line 115. I think we should be able to delete this entire `if` block and just have the `else` logic.

File: rubber_duck.py (hunk 5 above)
Comment (mmthornt) [reply rubber_duck.py]: This is the one thing I couldn't fix as it would struggle to define Attachment and fail

File: rubber_duck.py [hunk 9]
```diff
@@ -1,156 +1,165 @@
...
+    author_name: str
+    author_mention: str
+    message_id: int
+    content: str
+    file: list
+    
+class Attachment(TypedDict):
+    attachment_id: int
+    description: str
+    filename: str
```
Comment (byubean) [inline rubber_duck.py:36]: Define `Attachment` before `Message` so you can use the type for `file: list[Attachment]`

Comment (mmthornt) [reply rubber_duck.py:37]: This causes a NameError, even if I define Attachment before hand like this -> Attachment: dict    file: list[Attachment]
attempts at amending this have just led to more NameErrors even with further defintion

**Review by byubean — APPROVED** [review]

**Pull request merged**